compiler/src/iree/compiler/Dialect/Flow/Conversion/MeshToFlow/MeshToFlow.cpp

```diff
@@ -0,0 +1,768 @@
+
+#define GEN_PASS_CLASSES
+#define GEN_PASS_REGISTRATION
+#include "iree/compiler/Dialect/Flow/Conversion/MeshToFlow/Passes.h.inc" // IWYU pragma: keep
+
+// TODO: Use from MLIR when exposed.
+static mesh::ClusterOp getMesh(Operation *op, FlatSymbolRefAttr meshSymbol,
+                               SymbolTableCollection &symbolTableCollection) {
+  return symbolTableCollection.lookupNearestSymbolFrom<mesh::ClusterOp>(
+      op, meshSymbol);
+}
+
+template <typename MeshCollectiveOp>
+static mesh::ClusterOp getMesh(MeshCollectiveOp op,
+                               SymbolTableCollection &symbolTableCollection) {
+  return getMesh(op.getOperation(), op.getMeshAttr(), symbolTableCollection);
+}
+
+static bool hasMoreThanOneMesh(Operation *op) {
+  int meshCount = 0;
+  op->walk([&meshCount](mesh::ClusterOp mesh) {
+    ++meshCount;
+    return meshCount > 1 ? WalkResult::interrupt() : WalkResult::advance();
+  });
+  return meshCount > 1;
+}
+
+static bool isDefaultChannel(mesh::ClusterOp mesh,
+                             ArrayRef<mesh::MeshAxis> meshAxes) {
+  if (mesh.getRank() != static_cast<int64_t>(meshAxes.size())) {
+    return false;
+  }
+  mesh::MeshAxis i = 0;
+  return llvm::all_of(meshAxes, [&i](mesh::MeshAxis axis) {
+    bool res = (axis == i);
+    ++i;
+    return res;
+  });
+}
+
+static Value getDefaultChannel(mesh::ClusterOp mesh,
+                               bool useNamedDefaultChannels,
+                               ImplicitLocOpBuilder &builder) {
+  if (useNamedDefaultChannels)
+    return builder.create<IREE::Flow::ChannelDefaultOp>(mesh.getSymName());
+  else
+    return builder.create<IREE::Flow::ChannelDefaultOp>();
+}
+
+// Remove from `values` elements than have indices not present in filter.
+static SmallVector<Value> filterOutByIndex(ArrayRef<Value> values,
+                                           ArrayRef<mesh::MeshAxis> filter) {
+  SmallVector<Value> res;
+  for (size_t i = 0; i < values.size(); ++i) {
+    if (llvm::find(filter, i) == filter.end()) {
+      res.push_back(values[i]);
+    }
+  }
+  return res;
+}
+
+static CollectiveReductionOp convertReductionKind(mesh::Partial reduction) {
+  switch (reduction) {
+  case mesh::Partial::Max:
+    return CollectiveReductionOp::ReductionMaximum;
+  case mesh::Partial::Min:
+    return CollectiveReductionOp::ReductionMinimum;
+  case mesh::Partial::Sum:
+    return CollectiveReductionOp::ReductionSum;
+  default:
+    assert(false);
+    return CollectiveReductionOp::None;
+  }
+}
+
+static CollectiveReductionOpAttr
+convertReductionKind(mesh::PartialAttr reduction) {
+  return CollectiveReductionOpAttr::get(
+      reduction.getContext(), convertReductionKind(reduction.getValue()));
+}
+
+static Value buildChannelCreation(mesh::ClusterOp mesh,
+                                  ArrayRef<mesh::MeshAxis> meshAxes,
+                                  bool useNamedDefaultChannels,
+                                  ImplicitLocOpBuilder &builder) {
```

Comment:
Can we put some comment here to explain the logic? It's not trivially deducable; so better to be clear than requiring everybody looking at the code to rediscover it, especially regarding the channel split and key ordering. :)

Reply:
I added quite a bit of description in the `iree-convert-mesh-to-flow` pass description field. I think it is better to put the high-level description there in a more visible place.